src/celeritas/io/RootStepWriter.cc

```diff
@@ -0,0 +1,247 @@
+
+        // Store pre- and post-step
+        for (auto i : range(StepPoint::size_))
+        {
+            const auto& selection_point = selection_.points[i];
+            const auto& point           = steps.points[i];
+            auto&       tpoint          = tstep_.points[(int)i];
+
+            IF_POINT_SELECTED(volume_id, point.volume_id[tid].get());
+            IF_POINT_SELECTED(energy, point.energy[tid].value());
+            IF_POINT_SELECTED(time, point.time[tid]);
+            IF_POINT_REAL3_SELECTED(pos, tpoint.pos);
+            IF_POINT_REAL3_SELECTED(dir, tpoint.dir);
+        }
+
+        tstep_tree_->Fill();
+    }
+
+#undef IF_SELECTED
+#undef IF_POINT_SELECTED
+#undef IF_POINT_REAL3_SELECTED
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create steps tree. In order to have the option to individually select any
+ * member of `StepStateData` (StepData.hh) to be stored into the ROOT file, we
+ * cannot store an MC truth step object. This is accomplished by "flattening"
+ * the data so that each member of `TStepData` (MCTruthData.hh) is an
+ * individual branch that can be created based on the `StepSelection` booleans.
+ *
+ * To simplify the process of moving from Collection to a ROOT branch
+ * `TStepData` members *must* have the *exact* same name as the Collection
+ * members of `StepStateData`. This allows the use of C++ macros to
+ * automatically create ROOT branches with the same names as the struct
+ * elements.
+ */
+void RootStepWriter::make_tree()
+{
+#define CREATE_BRANCH_IF_SELECTED(ATTR)                     \
+    do                                                      \
+    {                                                       \
+        if (this->selection_.ATTR)                          \
+        {                                                   \
+            this->tstep_tree_->Branch(#ATTR, &tstep_.ATTR); \
+        }                                                   \
+    } while (0)
+
+#define CREATE_POINT_BRANCH_IF_SELECTED(ATTR, NAME, REF, TYPE)           \
+    do                                                                   \
+    {                                                                    \
+        if (this->selection_.ATTR)                                       \
+        {                                                                \
+            std::string leaf_def(NAME);                                  \
+            std::string str_attr(#ATTR);                                 \
+            const auto  last_p = str_attr.find_last_of(".");             \
+            if (str_attr.find("pos", last_p, 3) != std::string::npos     \
+                || str_attr.find("dir", last_p, 3) != std::string::npos) \
+            {                                                            \
+                leaf_def += "[3]";                                       \
+            }                                                            \
```

Comment:
This seems way too convoluted: it would be much safer and clearer to pass this separately for the arguments that need it (or use a little traits class or something to give the type string). But I don't understand why you need to pass the type explicitly at all since you don't do it for `CREATE_BRANCH_IF_SELECTED` ?

Also (even though these macros are used only locally) it's still good practice to namespace the macros.

Reply:
For arrays, the type can also be detected but only if you use `std::array` as the variable type.  (if you do so, then you can get rid of all the `leaf_def` code.

Reply:
then you can do
```
template <typename T>
TBranch *create_point_branch_if_selected(bool selection, const char *name, T &ref)
{
     if (selection)
        retrun this->tstep_tree_->Branch(name, ref); 
}
```
and
```
create_point_branch_if_selected( this->selection_.points[StepPoint::pre].volume_id,
   "points.pre.volume_id",
   tstep_.points[0].volume_id);
```
but at that point it might not even be worse the extra function.

PS.
An alternative that might work is to do:
```
auto branch = this->tstep_tree_->Branch("tstep.", tstep_);
if ( ! this->selection_.points[StepPoint::pre].volume_id )
   branch->GetListOfBranches()->Remove("tstep.points.pre.volume_id");
```
this gives a different layout in the `TTree` where the step object is preserved.
If it works as expected then setting up the reading would be trivial:
```
  this->tstep_tree_->SetBranchAddress("tstep.", &tstep_);
```
unconditionally (and of course only the member that were stored would be updated when calling `GetEntry`)

Reply:
Wow, so that way wouldn't take up the extra space for the omitted attributes? Sounds like the way to go for sure.